File: src/services/audit/detectors/ad/advanced/replication/duplicate-spn.ts
/**
 * Duplicate SPN Detector
 * Check for duplicate SPNs
 */

import { ADUser } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';

export function detectDuplicateSpn(users: ADUser[], includeDetails: boolean): Finding {
  const spnMap = new Map<string, string[]>();

  // Build SPN to user DN mapping
  users.forEach((u) => {
    const spns = (u as any).servicePrincipalName;
    if (spns && Array.isArray(spns)) {
      spns.forEach((spn: string) => {
        if (!spnMap.has(spn)) {
          spnMap.set(spn, []);
        }
        spnMap.get(spn)!.push(u.dn);
      });
    }
  });

  // Find duplicate SPNs
  const affected: string[] = [];
  spnMap.forEach((dns, _spn) => {
    if (dns.length > 1) {
      affected.push(...dns);
    }
  });

  return {
    type: 'DUPLICATE_SPN',
    severity: 'medium',
    category: 'advanced',
    title: 'Duplicate SPN',
    description: 'Service Principal Name registered multiple times. Can cause Kerberos authentication failures.',
    count: affected.length,
    affectedEntities: includeDetails ? affected : undefined,
  };
}

File: internal/audit/detectors/ad/advanced/replication/duplicate-spn.go
package replication

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// DuplicateSpnDetector detects duplicate SPNs
type DuplicateSpnDetector struct {
	audit.BaseDetector
}

// NewDuplicateSpnDetector creates a new detector
func NewDuplicateSpnDetector() *DuplicateSpnDetector {
	return &DuplicateSpnDetector{
		BaseDetector: audit.NewBaseDetector("DUPLICATE_SPN", audit.CategoryAdvanced),
	}
}

// Detect executes the detection
func (d *DuplicateSpnDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// Build SPN to DN mapping
	spnMap := make(map[string][]string)

	for _, u := range data.Users {
		for _, spn := range u.ServicePrincipalNames {
			spnMap[spn] = append(spnMap[spn], u.DN)
		}
	}

	for _, c := range data.Computers {
		for _, spn := range c.ServicePrincipalNames {
			spnMap[spn] = append(spnMap[spn], c.DN)
		}
	}

	// Find duplicate SPNs
	var affectedDNs []string
	for _, dns := range spnMap {
		if len(dns) > 1 {
			affectedDNs = append(affectedDNs, dns...)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Duplicate SPN",
		Description: "Service Principal Name registered multiple times. Can cause Kerberos authentication failures.",
		Count:       len(affectedDNs),
	}

	if data.IncludeDetails && len(affectedDNs) > 0 {
		entities := make([]types.AffectedEntity, len(affectedDNs))
		for i, dn := range affectedDNs {
			entities[i] = types.AffectedEntity{
				Type:           "object",
				SAMAccountName: dn,
			}
		}
		finding.AffectedEntities = entities
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewDuplicateSpnDetector())
}
